main/video.ts

```diff
@@ -0,0 +1,123 @@
+import path from 'path';
+import getFps from './utils/fps';
+import {getEncoding, convertToH264} from './utils/encoding';
+import {Rectangle, screen} from 'electron';
+import {Encoding} from './common/types';
+import {generateTimestampedName} from './utils/timestamped-name';
+
+interface VideoOptions {
+  filePath: string;
+  title?: string;
+  fps?: number;
+  encoding?: Encoding;
+  pixelDensity?: number;
+}
+
+export class Video {
+  static all = new Map<string, Video>();
+
+  static fromId(id: string) {
+    return this.all.get(id);
+  }
+
+  filePath: string;
+  title: string;
+  fps?: number;
+  encoding?: Encoding;
+  pixelDensity: number;
+  previewPath?: string;
+
+  isNewRecording = false;
+
+  isReady = false;
+  private readyPromise: Promise<void>;
+  private previewReadyPromise: Promise<string | undefined>;
+
+  constructor(options: VideoOptions) {
+    this.filePath = options.filePath;
+    this.title = options.title ?? path.basename(this.filePath);
+    this.fps = options.fps;
+    this.encoding = options.encoding;
+    this.pixelDensity = options.pixelDensity ?? 1;
+
+    Video.all.set(this.filePath, this);
+
+    this.readyPromise = this.collectInfo();
+    this.previewReadyPromise = this.readyPromise.then(() => this.getPreviewPath());
+  }
+
+  private async collectInfo() {
+    await Promise.all([
+      this.getFps(),
+      this.getEncoding(),
+    ]);
+
+    this.isReady = true;
+  }
+
+  async getFps() {
+    if (!this.fps) {
+      this.fps = Math.round(Number.parseFloat((await getFps(this.filePath)) ?? '0'));
+    }
+
+    return this.fps;
+  }
+
+  async getEncoding() {
+    if (!this.encoding) {
+      this.encoding = (await getEncoding(this.filePath)) as Encoding;
+    }
+
+    return this.encoding;
+  }
+
+  async getPreviewPath() {
+    if (!this.previewPath) {
+      if (this.encoding === 'h264') {
+        this.previewPath = this.filePath;
+      } else {
+        this.previewPath = await convertToH264(this.filePath);
```

Comment:
Written this like that (with a separate promise) as in the future when we have ProRes that takes a while to generate a preview, we can show the editor early with a spinner saying that the preview is being generated.